src/transforms/transaction.rs

```diff
@@ -0,0 +1,760 @@
+            .as_ref()
+            .map(|c| c.check(&event))
+            .unwrap_or(false);
+
+        let event = event.into_log();
+        let discriminant = Discriminant::from_log_event(&event, &self.identifier_fields);
+
+        if ends_here {
+            output.push(Event::from(
+                if let Some(mut state) = self.transaction_merge_states.remove(&discriminant) {
+                    state.add_event(event, &self.merge_strategies);
+                    state.flush()
+                } else {
+                    TransactionState::new(event, &self.merge_strategies).flush()
+                },
+            ));
+        } else {
+            match self.transaction_merge_states.entry(discriminant) {
+                hash_map::Entry::Vacant(entry) => {
+                    entry.insert(TransactionState::new(event, &self.merge_strategies));
+                }
+                hash_map::Entry::Occupied(mut entry) => {
+                    entry.get_mut().add_event(event, &self.merge_strategies);
+                }
+            }
+        }
+
+        self.flush_into(output);
+    }
+
+    fn transform_stream(
+        self: Box<Self>,
+        input_rx: Receiver<Event>,
+    ) -> Box<dyn Stream01<Item = Event, Error = ()> + Send>
+    where
+        Self: 'static,
+    {
+        let mut me = self;
+
+        let poll_period = me.flush_period.clone();
+        let mut last_flush = Instant::now();
+
+        let mut input_rx = Compat01As03::new(input_rx);
+
+        // let poller_shared_state = shared_state.clone();
+        let poll_flush = move |ctx: &mut Context| -> Poll<Option<Result<StreamEvent, ()>>> {
+            let p = Pin::new(&mut input_rx).poll_next(ctx);
+            match p {
+                Poll::Pending => {
+                    // If our input channel hasn't yielded anything
+                    let now = Instant::now();
+                    let since_last_flush = now.duration_since(last_flush);
+
+                    // But it hasn't been long enough to flush again yet.
+                    if let Some(flush_in) = poll_period.checked_sub(since_last_flush) {
+                        // Wake up when we're ready.
+                        let thread_waker = ctx.waker().clone();
+                        thread::spawn(move || {
+                            thread::sleep(flush_in);
+                            thread_waker.wake();
+                        });
+                        Poll::Pending
+                    } else {
+                        // Otherwise, it has been long enough since the last
+                        // flush to trigger another.
+                        last_flush = now;
+                        Poll::Ready(Some(Ok(StreamEvent::Flush)))
+                    }
+                }
+                // Pass `Poll<Option<Result<Event>>>` as `Poll<Option<Result<StreamEvent, ()>>>`
+                _ => return p.map(|p| p.map(|p| p.map(|p| StreamEvent::Event(p)))),
+            }
+        };
```

Comment:
We should be able to structure this in a simpler way. If I'm understanding correctly, we have two "triggers": incoming events and flush/expiration timers. Incoming events are already modeled as a stream, and we should be able to use a tokio `DelayQueue` as a stream of flush/expirations (with the additional benefit of keys, so we don't have to iterate and check our whole state in `flush_into`).

With both of those streams available, we should be able to "just" loop and `select` over them and do the appropriate flush/flush all/insert action directly (i.e. no need for the `StreamEvent` intermediary). We may need to use [`async-stream`](https://github.com/tokio-rs/async-stream) for that (/cc @LucioFranco) and then compat the resulting stream before returning it.

If that all works out (:crossed_fingers:), we can get rid of the thread spawn, most of the manual plumbing, the full state iteration, etc.

Reply:
I had a brief play around with creating a flush stream but I backed out as I couldn't see a simple and clean way of closing it once our event stream shuts down (without needing to return futures from `transform_stream`).

Not sure I'll have time this week to experiment.

Reply:
Not sure if this is useful or not but @MOZGIII has been working on an expiring type of hashmap for the new file sink https://github.com/timberio/vector/blob/tokio-compat-file-sink/src/expiring_hash_map/mod.rs, this may be useful.